x/evm/keeper/state_transition.go

```diff
@@ -326,6 +326,11 @@ func (k *Keeper) ApplyMessageWithConfig(
 	leftoverGas := msg.GasLimit
 	sender := vm.AccountRef(msg.From)
 	tracer := cfg.GetTracer()
+	debugFn := func() {
+		if tracer != nil && cfg.DebugTrace {
+			stateDB.AddBalance(sender.Address(), new(big.Int).Mul(msg.GasPrice, new(big.Int).SetUint64(leftoverGas)))
+		}
+	}
```

Comment:
I feel that we should have a better structure of the code in this core function, I think the issue is introduced in #545, right? should we try step back and refactor the code in a better way?

Reply:
balance seems not match by simple revert it

Reply:
There's [refund](https://github.com/crypto-org-chain/ethermint/blob/b602a262af79ae9cad244e8a73f0a80d701d15ee/x/evm/keeper/state_transition.go#L227) after ApplyMessageWithConfig when tx gets excuted, but balance gets restored after commit when debug trace that tx, and [AddBalance](https://github.com/crypto-org-chain/ethermint/blob/b602a262af79ae9cad244e8a73f0a80d701d15ee/x/evm/keeper/state_transition.go#L341) is based on previous value.